internal/server/run.go

```diff
@@ -0,0 +1,92 @@
+package server
+
+import (
+	"errors"
+	"io"
+	"log"
+	"net"
+	"tcp-load-balancer/internal/upstream"
+)
+
+var ErrUninitialized = errors.New("load balancer not initialized")
+
+// Run handles incoming connections until terminated.
+func (l *LoadBalancer) Run() error {
+	if l == nil || l.listener == nil {
+		return ErrUninitialized
+	}
+
+	for {
+		// TODO: Set up mTLS in next PR. For now, connect without TLS.
+		clientConn, err := l.listener.Accept()
+		if err != nil {
+			log.Printf("Load balancer unexpectedly declined connection and will be shut down: %s", err)
+			// TODO: attempt to re-establish the listener with a retry mechanism (leaving out of scope for this project).
+			break
+		}
+
+		if err := l.handleConnection(clientConn); err != nil {
+			log.Printf("Unable to handle connection: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// handleConnection selects an upstream host, tracks connection counts, and forwards data upstream.
+func (l *LoadBalancer) handleConnection(clientConn net.Conn) error {
+	// Host selection is not included in goroutine handling so that requests arriving at the same time are not routed to the same host.
+	// This adds a small amount of latency to the request, but ensures accurate load balancing.
+	host, err := l.LeastConnections()
```

Comment:
Rather than using the mutex, `handleConnection` (and also `LeastConnections`) is always called from the main goroutine, and never used concurrently. Same with `IncrementActiveConnections.` See comment on line 38. 

So the current flow is:
```
// l.listener.Accept()
// handleConnection
// - select host from LeastConnections
// - IncrementActiveConnections
// - start goroutine
// - (concurrent) forwardToHost
// - (concurrent) DecrementActiveConnections
```

I was thinking that running these steps from the main goroutine would be equivalent to having the LeastConnections and IncrementActiveConnections locked behind a mutex.